Tools/Scripts/webkitpy/benchmark_runner/github_downloader.py

```diff
@@ -0,0 +1,195 @@
+import logging
+import os.path
+import re
+import subprocess
+import time
+from multiprocessing import Pool
+from functools import wraps
+
+import requests
+
+
+_log = logging.getLogger(__name__)
+
+
+class GithubDownloadTask(object):
+    URL_RE = re.compile(r'\Ahttps?://github.(?P<domain>\S+)/(?P<owner>\S+)/(?P<repository>\S+)/tree/(?P<commit>[0-9a-zA-Z]+)/(?P<relpath>\S+)\Z')
+
+    def __init__(self, source_url, subtree=None):
+        match = self.URL_RE.match(source_url)
+        assert match, f'Unrecognized source URL: {source_url}'
+        owner = match.group('owner')
+        repository = match.group('repository')
+        api_url = 'https://api.github.{}'.format(match.group('domain'))
+        self._downloader = GithubDownloader(api_url, owner, repository)
+        self._commit = match.group('commit')
+        self._relpath = match.group('relpath')
+        self._subtree = subtree
+
+    def execute(self, dest):
+        self._downloader.download_content(self._relpath, self._commit, dest, subtree=self.subtree)
+
+    @property
+    def subtree(self):
+        if self._subtree is None:
+            self._subtree = self._downloader.subtree_for_path_with_commit(self._relpath, self._commit)
+        return self._subtree
+
+
+def retry_on_exception(count):
+    def decorator(function):
+        @wraps(function)
+        def wrapped(*args, **kwargs):
+            remaining_retry = count + 1
+            while remaining_retry:
+                remaining_retry -= 1
+                try:
+                    return function(*args, **kwargs)
+                except Exception as error:
+                    _log.warning(f'Encountered {error} in {function.__name__}, remaining retry: {remaining_retry}.')
+                    if not remaining_retry:
+                        raise
+        return wrapped
+    return decorator
+
+
+class GithubDownloader(object):
+    def __init__(self, url, owner, repository):
+        self._url = url
+        self._owner = owner
+        self._repository = repository
+        self._session = None
+
+    @classmethod
+    def is_directory(cls, entry):
+        return entry['mode'].startswith('040')
```

Comment:
why can't we use return entry['type'] == 'tree'?

Reply:
I guess we can, using `mode` is equivalent according to git documentation.

Reply:
Updated using `entry['type'] == 'tree'`.